File: crates/deepsec-processor/src/revalidate.rs
use crate::agents::{AgentBackend, RevalidateInput, RevalidateInputFinding};
use crate::errors::ProcessorError;
use deepsec_core::ids::{generate_run_id, now_iso};
use deepsec_core::store::{load_all_file_records, write_file_record};
use deepsec_core::{
    DataRoot, InvocationMode, ProcessorConfig, Revalidation, RevalidationVerdict, RunPhase,
    RunType, Severity, complete_run, create_run_meta, write_run_meta,
};
use indexmap::IndexMap;
use std::path::PathBuf;

pub struct RevalidateOptions {
    pub project_id: String,
    pub project_root: PathBuf,
    pub data_root: DataRoot,
    pub backend: std::sync::Arc<dyn AgentBackend>,
    pub filter_prefix: Option<String>,
    pub force: bool,
}

#[derive(Debug, Default, Clone)]
pub struct RevalidateOutcome {
    pub run_id: String,
    pub revalidated: usize,
    pub true_positives: usize,
    pub false_positives: usize,
    pub fixed: usize,
    pub uncertain: usize,
}

pub async fn run_revalidate(opts: RevalidateOptions) -> Result<RevalidateOutcome, ProcessorError> {
    let run_id = generate_run_id();
    let mut meta = create_run_meta(
        &opts.project_id,
        &run_id,
        opts.project_root.to_string_lossy().as_ref(),
        RunType::Revalidate,
    );
    meta.processor_config = Some(ProcessorConfig {
        agent_type: opts.backend.kind().as_str().into(),
        model: opts.backend.model().into(),
        model_config: IndexMap::new(),
        invocation_mode: Some(InvocationMode::Scan),
        source: None,
    });
    write_run_meta(&opts.data_root, &meta)?;

    let records = load_all_file_records(&opts.data_root, &opts.project_id)
        .map_err(|e| ProcessorError::Other(anyhow::Error::new(e)))?;

    let mut outcome = RevalidateOutcome {
        run_id: run_id.clone(),
        ..Default::default()
    };

    for mut rec in records {
        if let Some(prefix) = &opts.filter_prefix {
            if !rec.file_path.starts_with(prefix) {
                continue;
            }
        }
        let needs: Vec<(usize, &deepsec_core::Finding)> = rec
            .findings
            .iter()
            .enumerate()
            .filter(|(_, f)| opts.force || f.revalidation.is_none())
            .collect();
        if needs.is_empty() {
            continue;
        }
        let inputs: Vec<RevalidateInputFinding> = needs
            .iter()
            .map(|(i, f)| RevalidateInputFinding {
                index: *i,
                severity: f.severity,
                vuln_slug: f.vuln_slug.clone(),
                title: f.title.clone(),
                description: f.description.clone(),
                line_numbers: f.line_numbers.clone(),
            })
            .collect();
        let content = fs_err::read_to_string(opts.project_root.join(&rec.file_path))
            .unwrap_or_default();
        let input = RevalidateInput {
            project_root: opts.project_root.clone(),
            file_path: rec.file_path.clone(),
            file_content: content,
            findings: inputs,
        };
        match opts.backend.revalidate(&input).await {
            Ok((revs, _usage, _dur)) => {
                for r in revs {
                    let verdict = match r.verdict.as_str() {
                        "true-positive" => Some(RevalidationVerdict::TruePositive),
                        "false-positive" => Some(RevalidationVerdict::FalsePositive),
                        "fixed" => Some(RevalidationVerdict::Fixed),
                        "uncertain" => Some(RevalidationVerdict::Uncertain),
                        "accepted-risk" => Some(RevalidationVerdict::AcceptedRisk),
                        _ => None,
                    };
                    let Some(v) = verdict else { continue };
                    if let Some(f) = rec.findings.get_mut(r.index) {
                        f.revalidation = Some(Revalidation {
                            verdict: v,
                            reasoning: r.reasoning,
                            adjusted_severity: r.adjusted_severity.map(Severity::from_),
                            revalidated_at: now_iso(),
                            run_id: run_id.clone(),
                            model: opts.backend.model().into(),
                        });
                        outcome.revalidated += 1;
                        match v {
                            RevalidationVerdict::TruePositive => outcome.true_positives += 1,
                            RevalidationVerdict::FalsePositive => outcome.false_positives += 1,
                            RevalidationVerdict::Fixed => outcome.fixed += 1,
                            RevalidationVerdict::Uncertain => outcome.uncertain += 1,
                            _ => {}
                        }
                    }
                }
                write_file_record(&opts.data_root, &rec)
                    .map_err(|e| ProcessorError::Other(anyhow::Error::new(e)))?;
            }
            Err(ProcessorError::Quota(q)) => {
                tracing::warn!("quota exhausted: {} {}", q.backend, q.detail);
                break;
            }
            Err(e) => {
                tracing::error!("revalidate failed for {}: {e}", rec.file_path);
            }
        }
    }

    meta.stats.findings_revalidated = Some(outcome.revalidated);
    meta.stats.true_positives = Some(outcome.true_positives);
    meta.stats.false_positives = Some(outcome.false_positives);
    meta.stats.fixed = Some(outcome.fixed);
    meta.stats.uncertain = Some(outcome.uncertain);
    write_run_meta(&opts.data_root, &meta)?;
    complete_run(&opts.data_root, &opts.project_id, &run_id, RunPhase::Done)
        .map_err(|e| ProcessorError::Other(anyhow::Error::new(e)))?;
    Ok(outcome)
}

// Trampoline so we can call `Severity::from_("HIGH")` style.
trait SevFromExt {
    fn from_(s: Severity) -> Severity;
}
impl SevFromExt for Severity {
    fn from_(s: Severity) -> Severity {
        s
    }
}

File: internal/processor/revalidate.go
package processor

import (
	"context"
	"os"
	"path/filepath"

	"github.com/noeljackson/deepsec/internal/core"
)

// RevalidateOptions configures a revalidate run.
type RevalidateOptions struct {
	ProjectID    string
	ProjectRoot  string
	DataRoot     core.DataRoot
	Backend      AgentBackend
	ProviderName string
	FilterPrefix string
	Force        bool
}

// RevalidateOutcome summarizes the run.
type RevalidateOutcome struct {
	RunID          string
	Revalidated    int
	TruePositives  int
	FalsePositives int
	Fixed          int
	Uncertain      int
}

// Revalidate re-checks every finding without a revalidation field (or
// every finding, if Force is set) against the current source.
func Revalidate(ctx context.Context, opts RevalidateOptions) (*RevalidateOutcome, error) {
	runID := core.GenerateRunID()
	meta := core.NewRunMeta(opts.ProjectID, runID, opts.ProjectRoot, core.RunTypeRevalidate)
	meta.ProcessorConfig = &core.ProcessorConfig{
		AgentType:      opts.ProviderName,
		Model:          opts.Backend.Model(),
		ModelConfig:    map[string]any{},
		InvocationMode: core.InvocationModeScan,
	}
	if err := opts.DataRoot.WriteRunMeta(meta); err != nil {
		return nil, err
	}

	records, err := opts.DataRoot.LoadAllFileRecords(opts.ProjectID)
	if err != nil {
		return nil, err
	}
	outcome := &RevalidateOutcome{RunID: runID}

	for _, rec := range records {
		if opts.FilterPrefix != "" && !hasPrefix(rec.FilePath, opts.FilterPrefix) {
			continue
		}
		var inputs []RevalidateInputFinding
		for i, f := range rec.Findings {
			if !opts.Force && f.Revalidation != nil {
				continue
			}
			inputs = append(inputs, RevalidateInputFinding{
				Index:       i,
				Severity:    f.Severity,
				VulnSlug:    f.VulnSlug,
				Title:       f.Title,
				Description: f.Description,
				LineNumbers: f.LineNumbers,
			})
		}
		if len(inputs) == 0 {
			continue
		}
		body, _ := os.ReadFile(filepath.Join(opts.ProjectRoot, rec.FilePath))
		in := &RevalidateInput{
			ProjectRoot: opts.ProjectRoot,
			FilePath:    rec.FilePath,
			FileContent: string(body),
			Findings:    inputs,
		}
		revs, _, _, err := opts.Backend.Revalidate(ctx, in)
		if err != nil {
			if IsQuotaErr(err) {
				break
			}
			continue
		}
		for _, r := range revs {
			if r.Index < 0 || r.Index >= len(rec.Findings) {
				continue
			}
			f := &rec.Findings[r.Index]
			f.Revalidation = &core.Revalidation{
				Verdict:          r.Verdict,
				Reasoning:        r.Reasoning,
				AdjustedSeverity: r.AdjustedSeverity,
				RevalidatedAt:    core.NowISO(),
				RunID:            runID,
				Model:            opts.Backend.Model(),
			}
			outcome.Revalidated++
			switch r.Verdict {
			case core.VerdictTruePositive, core.VerdictAcceptedRisk:
				outcome.TruePositives++
			case core.VerdictFalsePositive:
				outcome.FalsePositives++
			case core.VerdictFixed:
				outcome.Fixed++
			case core.VerdictUncertain:
				outcome.Uncertain++
			}
		}
		if err := opts.DataRoot.WriteFileRecord(rec); err != nil {
			return nil, err
		}
	}

	meta.Stats.FindingsRevalidated = &outcome.Revalidated
	meta.Stats.TruePositives = &outcome.TruePositives
	meta.Stats.FalsePositives = &outcome.FalsePositives
	meta.Stats.Fixed = &outcome.Fixed
	meta.Stats.Uncertain = &outcome.Uncertain
	if err := opts.DataRoot.WriteRunMeta(meta); err != nil {
		return nil, err
	}
	if _, err := opts.DataRoot.CompleteRun(opts.ProjectID, runID, core.RunPhaseDone); err != nil {
		return nil, err
	}
	return outcome, nil
}
